### PR TITLE
New Candidate+Point seeded tracking region producer for Tau HLT tracking

### PR DESCRIPTION
**Pull request by tstreble**
Add a new tracking region producer seeded by candidates + points for Tau HLT Tracking.
To be used for triplet+doublet recovery of HLT paths with regional PF tracking.

### CODE REVIEW

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTracker/TkTrackingRegions/plugins/CandidatePointSeededTrackingRegionsProducer.h [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+	double deltaPhi_RoI = deltaPhi(phiMax_RoI,phi_RoI);
+
+       	double x = std::cos(phi_RoI);
+	double y = std::sin(phi_RoI);
+	double theta = 2*std::atan(std::exp(-eta_RoI));
+	double z = 1./std::tan(theta);
+
+	GlobalVector direction( x,y,z );
+      
+	for (size_t  j=0; j<origins.size() && n_regions < m_maxNRegions; ++j)
```
Comment (makortel) [inline RecoTracker/TkTrackingRegions/plugins/CandidatePointSeededTrackingRegionsProducer.h]: Loop variable better be `k` or something else to not clash with `j` or `i` of the inner loops (or convert to range-for).

Comment (makortel) [reply RecoTracker/TkTrackingRegions/plugins/CandidatePointSeededTrackingRegionsProducer.h]: Even if in practice this appears to be safe (the `j` in inner block refers to the `j` declared in the inner block), but to not confuse the reader.

Comment (tstreble) [reply RecoTracker/TkTrackingRegions/plugins/CandidatePointSeededTrackingRegionsProducer.h]: Definitely makes sense
Thanks for spotting this

**Comment by perrotta** [comment]
please test

**Comment by makortel** [comment]
Replying to @slava77 (https://github.com/cms-sw/cmssw/pull/20371#discussion_r137576805) in the master PR. In general the PR looks fine to me, and I collected some remarks below (going also more general from this PR):

- I like the approach to combine the candidate-seeded and "point"-seeded regions in a single producer
- There is some copy-paste between `CandidateSeeded`, `PointSeeded`, and the new `CandidatePointSeeded` producers, but I suggest we deal with that later (I assume the new producer is wanted online rather sooner than later)
- I welcome the move of the TrackingRegionProducers under `RecoTracker/TkTrackingRegions`
  * I wouldn't mind if the rest in `RecoTauTag/HLTProducers` would be moved as well (not necessarily in this PR though)
- We should probably look to clean up the old `TrackingRegionProducerFactory`

**Comment by slava77** [comment]
On 9/8/17 5:08 AM, Matti Kortelainen wrote:
> Replying to @slava77 <https://github.com/slava77> (#20371 (comment)
> <https://github.com/cms-sw/cmssw/pull/20371#discussion_r137576805>) in
> the master PR. In general the PR looks fine to me, and I collected some
> remarks below (going also more general from this PR):
> 
>   * I like the approach to combine the candidate-seeded and
>     "point"-seeded regions in a single producer
>   * There is some copy-paste between |CandidateSeeded|, |PointSeeded|,
>     and the new |CandidatePointSeeded| producers, but I suggest we deal
>     with that later (I assume the new producer is wanted online rather
>     sooner than later)



How significant is this copy-paste?
If it's a matter of ~50 lines in common, OK.
If it's essentially the full file with only ~50 files differing, better
merge now.

**Comment by makortel** [comment]
> How significant is this copy-paste?
> If it's a matter of ~50 lines in common, OK.
> If it's essentially the full file with only ~50 files differing, better
> merge now.

Taking 3-way diff and applying some heuristic which lines to count, I'd say there is ~90-100 lines in common with all three.

**Comment by Martin-Grunewald** [comment]
+1

**Comment by perrotta** [comment]
Indeed the code of the three producers is almost identical except in the part of the nested loops inside the methods Regions(...) labelled as
"create tracking regions (maximum MaxNRegions of them) in directions of the objects of interest"

If I am not wrong, all other differences are in the definition of the different parameters needed there, plus some different indentation and so on.

Please verify (and tell me if I'm wrong), but my impression is that the three codes can be easily merged.

**Comment by tstreble** [comment]
In principle, we could imagine a single implementation of this producer with a string input parameter specifying which version should be used (candidate, point or candidate+point seeded).
The only downside of this is that it will require a synchronized update of all the existing tracking sequences currently using the candidate- or the point-seeded tracking region producers.
As two different producers are already in use, I think it will not be possible to define a default behavior which would leave unchanged the existing HLT configuration.
@perrotta @silviodonato @Martin-Grunewald @JanFSchulte Do you have any thoughts on this?

**Comment by JanFSchulte** [comment]
The overwhelming majority of sequences use the CandidateSeeded producer, with only a handful of cases using PointSeeded. If CandidateSeeded would become the default behaviour, it might be feasible to merge all producers with moderate effort. But as I imagine that TAU wants to get this improvement to their sequences online ASAP I would be in favor of merging this PR as is and consolidate later. 

Cheers, 
Jan

**Comment by fwyzard** [comment]
"consolidate later" has a tendency of becoming "never" way too often...

**Comment by JanFSchulte** [comment]
If delyaing a bit is fine for TAU, I'm all for it of course.

**Comment by makortel** [comment]
In the long term it would be clearest to me if the combined "candidate and/or point seeded" producer would be called `CandidatePointSeededTrackingRegionsProducer` or similar (like in this PR).

Said that, for me it is fine to make a "detour" via `PointSeededTrackingRegionsProducer` if it makes it easier to backport to 92X and update the HLT menu there. But eventually I'd like to see the class renamed and all uses of `CandidateSeededTrackingRegionsProducer` migrated to the new producer.

**Comment by tstreble** [comment]
I'm working on an implementation making everything available in `CandidatePointSeededTrackingRegionsProducer`.
Ideally, we would have as a temporary solution the three producers available, such that HLT operations is not disturbed.
Once the new producer is integrated (with the three possible variations implemented), HLT configuration can be updated and it will be possible to deprecate the existing `CandidateSeededTrackingRegionsProducer` and `PointSeededTrackingRegionsProducer`.
I hope this solution is fine for everyone.

**Comment by Martin-Grunewald** [comment]
Could you then please close this PR and also #20371 for now?

**Comment by tstreble** [comment]
Ok for me

**Comment by perrotta** [comment]
@tstreble : your plan looks fine, thank you.
Do you have a time estimate for it?

**Comment by tstreble** [comment]
I hope to have something ready by the end of the week.

**Comment by perrotta** [comment]
Better soonish...

https://twiki.cern.ch/twiki/bin/viewauth/CMS/CMSSW_9_4_0


tstreble <notifications@github.com> ha scritto:

**Comment by tstreble** [comment]
Superseded by #20672

**Pull request closed without merging**